### ISSUE TITLE
TestPeerConnection_RoleSwitch fails

### ISSUE THREAD

**Issue by at-wat** [pion/webrtc#1074]
https://travis-ci.org/pion/webrtc/jobs/660184903
It fails again on rebuild.
```
=== RUN   TestPeerConnection_RoleSwitch
goroutine profile: total 30
4 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82bb142 0x83006fc 0x82c07f9 0x809e321
#	0x82bb141	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.8.10/packetio/buffer.go:135
#	0x83006fb	github.com/pion/webrtc/v2/internal/mux.(*Endpoint).Read+0x3b	/go/src/github.com/pion/webrtc/internal/mux/endpoint.go:36
#	0x82c07f8	github.com/pion/srtp.(*session).start.func1+0x98		/go/pkg/mod/github.com/pion/srtp@v1.2.7/session.go:124

4 @ 0x80758a4 0x8085981 0x808595c 0x80a5d66 0x8214719 0x8218430 0x83f0d48 0x809e321
#	0x808595b	sync.runtime_notifyListWait+0xfb				/usr/local/go/src/runtime/sema.go:510
#	0x80a5d65	sync.(*Cond).Wait+0x65						/usr/local/go/src/sync/cond.go:56
#	0x8214718	github.com/pion/sctp.(*Stream).ReadSCTP+0x88			/go/pkg/mod/github.com/pion/sctp@v1.7.6/stream.go:107
#	0x821842f	github.com/pion/datachannel.(*DataChannel).ReadDataChannel+0x3f	/go/pkg/mod/github.com/pion/datachannel@v1.4.16/datachannel.go:188
#	0x83f0d47	github.com/pion/webrtc/v2.(*DataChannel).readLoop+0x87		/go/src/github.com/pion/webrtc/datachannel.go:319

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82bb142 0x82f5705 0x8300e6f 0x809e321
#	0x82bb141	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.8.10/packetio/buffer.go:135
#	0x82f5704	github.com/pion/ice.(*Conn).Read+0x74				/go/pkg/mod/github.com/pion/ice@v0.7.10/transport.go:78
#	0x8300e6e	github.com/pion/webrtc/v2/internal/mux.(*Mux).readLoop+0x9e	/go/src/github.com/pion/webrtc/internal/mux/mux.go:106

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82bb142 0x83006fc 0x829aed9 0x809e321
#	0x82bb141	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.8.10/packetio/buffer.go:135
#	0x83006fb	github.com/pion/webrtc/v2/internal/mux.(*Endpoint).Read+0x3b	/go/src/github.com/pion/webrtc/internal/mux/endpoint.go:36
#	0x829aed8	github.com/pion/dtls/v2.(*Conn).inboundLoop+0x98		/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.7/conn.go:583

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c77c 0x821800b 0x8217fea 0x84045e5 0x809e321
#	0x821800a	github.com/pion/sctp.(*Association).AcceptStream+0x3a			/go/pkg/mod/github.com/pion/sctp@v1.7.6/association.go:1123
#	0x8217fe9	github.com/pion/datachannel.Accept+0x19					/go/pkg/mod/github.com/pion/datachannel@v1.4.16/datachannel.go:129
#	0x84045e4	github.com/pion/webrtc/v2.(*SCTPTransport).acceptDataChannels+0x64	/go/src/github.com/pion/webrtc/sctptransport.go:143

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c77c 0x82bd3c0 0x8432dd7 0x809e321
#	0x82bd3bf	github.com/pion/srtp.(*SessionSRTCP).AcceptStream+0x3f			/go/pkg/mod/github.com/pion/srtp@v1.2.7/session_srtcp.go:76
#	0x8432dd6	github.com/pion/webrtc/v2.(*PeerConnection).drainSRTP.func3+0x86	/go/src/github.com/pion/webrtc/peerconnection.go:980

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c77c 0x82be030 0x8432b57 0x809e321
#	0x82be02f	github.com/pion/srtp.(*SessionSRTP).AcceptStream+0x3f			/go/pkg/mod/github.com/pion/srtp@v1.2.7/session_srtp.go:77
#	0x8432b56	github.com/pion/webrtc/v2.(*PeerConnection).drainSRTP.func2+0x46	/go/src/github.com/pion/webrtc/peerconnection.go:960

2 @ 0x80758a4 0x8070524 0x806f9fb 0x80be807 0x80bf396 0x80bf37b 0x81bb7ef 0x81cf63f 0x81ce407 0x82ede10 0x809e321
#	0x806f9fa	internal/poll.runtime_pollWait+0x4a			/usr/local/go/src/runtime/netpoll.go:184
#	0x80be806	internal/poll.(*pollDesc).wait+0x36			/usr/local/go/src/internal/poll/fd_poll_runtime.go:87
#	0x80bf395	internal/poll.(*pollDesc).waitRead+0x165		/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
#	0x80bf37a	internal/poll.(*FD).ReadFrom+0x14a			/usr/local/go/src/internal/poll/fd_unix.go:219
#	0x81bb7ee	net.(*netFD).readFrom+0x3e				/usr/local/go/src/net/fd_unix.go:208
#	0x81cf63e	net.(*UDPConn).readFrom+0x3e				/usr/local/go/src/net/udpsock_posix.go:47
#	0x81ce406	net.(*UDPConn).ReadFrom+0x46				/usr/local/go/src/net/udpsock.go:121
#	0x82ede0f	github.com/pion/ice.(*candidateBase).recvLoop+0xef	/go/pkg/mod/github.com/pion/ice@v0.7.10/candidate_base.go:92

2 @ 0x80758a4 0x8083e59 0x81fb413 0x809e321
#	0x81fb412	github.com/pion/sctp.(*Association).writeLoop+0x122	/go/pkg/mod/github.com/pion/sctp@v1.7.6/association.go:456

2 @ 0x80758a4 0x8083e59 0x829901f 0x81fb13c 0x809e321
#	0x829901e	github.com/pion/dtls/v2.(*Conn).Read+0xae		/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.7/conn.go:285
#	0x81fb13b	github.com/pion/sctp.(*Association).readLoop+0x13b	/go/pkg/mod/github.com/pion/sctp@v1.7.6/association.go:421

2 @ 0x80758a4 0x8083e59 0x82f854f 0x809e321
#	0x82f854e	github.com/pion/ice.(*Agent).startConnectivityChecks.func1.1+0xae	/go/pkg/mod/github.com/pion/ice@v0.7.10/agent.go:612

1 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x8110f5b 0x8113daa 0x8110c2a 0x8111fcd 0x811127b 0x844de44 0x80754d2 0x809e321
#	0x8110f5a	testing.(*T).Run+0x2ca		/usr/local/go/src/testing/testing.go:961
#	0x8113da9	testing.runTests.func1+0x59	/usr/local/go/src/testing/testing.go:1202
#	0x8110c29	testing.tRunner+0x99		/usr/local/go/src/testing/testing.go:909
#	0x8111fcc	testing.runTests+0x22c		/usr/local/go/src/testing/testing.go:1200
#	0x811127a	testing.(*M).Run+0x13a		/usr/local/go/src/testing/testing.go:1117
#	0x844de43	main.main+0x103			_testmain.go:322
#	0x80754d1	runtime.main+0x201		/usr/local/go/src/runtime/proc.go:203

1 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82bb142 0x8402cc6 0x8402c9d 0x840a377 0x840a3ff 0x840a985 0x83fc0ba 0x809e321
#	0x82bb141	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.8.10/packetio/buffer.go:135
#	0x8402cc5	github.com/pion/srtp.(*ReadStreamSRTP).Read+0x55		/go/pkg/mod/github.com/pion/srtp@v1.2.7/stream_srtp.go:69
#	0x8402c9c	github.com/pion/webrtc/v2.(*RTPReceiver).readRTP+0x2c		/go/src/github.com/pion/webrtc/rtpreceiver.go:155
#	0x840a376	github.com/pion/webrtc/v2.(*Track).Read+0xb6			/go/src/github.com/pion/webrtc/track.go:97
#	0x840a3fe	github.com/pion/webrtc/v2.(*Track).ReadRTP+0x5e			/go/src/github.com/pion/webrtc/track.go:103
#	0x840a984	github.com/pion/webrtc/v2.(*Track).determinePayloadType+0x34	/go/src/github.com/pion/webrtc/track.go:198
#	0x83fc0b9	github.com/pion/webrtc/v2.(*PeerConnection).startReceiver+0xe9	/go/src/github.com/pion/webrtc/peerconnection.go:784

1 @ 0x80758a4 0x8083e59 0x8420f6a 0x84221da 0x8110c2a 0x809e321
#	0x8420f69	github.com/pion/webrtc/v2.sendVideoUntilDone+0x129		/go/src/github.com/pion/webrtc/peerconnection_renegotation_test.go:19
#	0x84221d9	github.com/pion/webrtc/v2.TestPeerConnection_RoleSwitch+0x4d9	/go/src/github.com/pion/webrtc/peerconnection_renegotation_test.go:195
#	0x8110c29	testing.tRunner+0x99						/usr/local/go/src/testing/testing.go:909

1 @ 0x814626c 0x81460da 0x8143133 0x83ee9d7 0x809e321
#	0x814626b	runtime/pprof.writeRuntimeProfile+0x7b			/usr/local/go/src/runtime/pprof/pprof.go:708
#	0x81460d9	runtime/pprof.writeGoroutine+0x79			/usr/local/go/src/runtime/pprof/pprof.go:670
#	0x8143132	runtime/pprof.(*Profile).WriteTo+0x2c2			/usr/local/go/src/runtime/pprof/pprof.go:329
#	0x83ee9d6	github.com/pion/transport/test.TimeOut.func1+0x56	/go/pkg/mod/github.com/pion/transport@v0.8.10/test/util.go:18

panic: timeout

goroutine 1695 [running]:
github.com/pion/transport/test.TimeOut.func1()
	/go/pkg/mod/github.com/pion/transport@v0.8.10/test/util.go:21 +0xda
created by time.goFunc
	/usr/local/go/src/time/sleep.go:168 +0x37
FAIL	github.com/pion/webrtc/v2	87.472s
```

**Comment by at-wat** [comment]
It also fails on i386 Go 1.12.
Seems not Go 1.13 specific.
https://travis-ci.org/pion/webrtc/jobs/660251538

**Comment by Sean-Der** [comment]
So frustrating. It landed and was fine for a few days, and just started flaking -_-

Will dig into it @at-wat !

**Comment by at-wat** [comment]
It still fails sometimes.
https://travis-ci.org/github/pion/webrtc/jobs/664710463 (restarted)
```
=== RUN   TestPeerConnection_RoleSwitch
goroutine profile: total 29
4 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82bb1a2 0x830075c 0x82c0859 0x809e321
#	0x82bb1a1	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.9.2/packetio/buffer.go:136
#	0x830075b	github.com/pion/webrtc/v2/internal/mux.(*Endpoint).Read+0x3b	/go/src/github.com/pion/webrtc/internal/mux/endpoint.go:36
#	0x82c0858	github.com/pion/srtp.(*session).start.func1+0x98		/go/pkg/mod/github.com/pion/srtp@v1.2.7/session.go:124

4 @ 0x80758a4 0x8085981 0x808595c 0x80a5d66 0x8214779 0x8218490 0x83f0db8 0x809e321
#	0x808595b	sync.runtime_notifyListWait+0xfb				/usr/local/go/src/runtime/sema.go:510
#	0x80a5d65	sync.(*Cond).Wait+0x65						/usr/local/go/src/sync/cond.go:56
#	0x8214778	github.com/pion/sctp.(*Stream).ReadSCTP+0x88			/go/pkg/mod/github.com/pion/sctp@v1.7.6/stream.go:107
#	0x821848f	github.com/pion/datachannel.(*DataChannel).ReadDataChannel+0x3f	/go/pkg/mod/github.com/pion/datachannel@v1.4.16/datachannel.go:188
#	0x83f0db7	github.com/pion/webrtc/v2.(*DataChannel).readLoop+0x87		/go/src/github.com/pion/webrtc/datachannel.go:319

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82bb1a2 0x82f5765 0x8300ecf 0x809e321
#	0x82bb1a1	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.9.2/packetio/buffer.go:136
#	0x82f5764	github.com/pion/ice.(*Conn).Read+0x74				/go/pkg/mod/github.com/pion/ice@v0.7.10/transport.go:78
#	0x8300ece	github.com/pion/webrtc/v2/internal/mux.(*Mux).readLoop+0x9e	/go/src/github.com/pion/webrtc/internal/mux/mux.go:106

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x82bb1a2 0x830075c 0x829af39 0x809e321
#	0x82bb1a1	github.com/pion/transport/packetio.(*Buffer).Read+0x41		/go/pkg/mod/github.com/pion/transport@v0.9.2/packetio/buffer.go:136
#	0x830075b	github.com/pion/webrtc/v2/internal/mux.(*Endpoint).Read+0x3b	/go/src/github.com/pion/webrtc/internal/mux/endpoint.go:36
#	0x829af38	github.com/pion/dtls/v2.(*Conn).inboundLoop+0x98		/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.7/conn.go:583

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c77c 0x821806b 0x821804a 0x84045f5 0x809e321
#	0x821806a	github.com/pion/sctp.(*Association).AcceptStream+0x3a			/go/pkg/mod/github.com/pion/sctp@v1.7.6/association.go:1123
#	0x8218049	github.com/pion/datachannel.Accept+0x19					/go/pkg/mod/github.com/pion/datachannel@v1.4.16/datachannel.go:129
#	0x84045f4	github.com/pion/webrtc/v2.(*SCTPTransport).acceptDataChannels+0x64	/go/src/github.com/pion/webrtc/sctptransport.go:143

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c77c 0x82bd420 0x8433047 0x809e321
#	0x82bd41f	github.com/pion/srtp.(*SessionSRTCP).AcceptStream+0x3f			/go/pkg/mod/github.com/pion/srtp@v1.2.7/session_srtcp.go:76
#	0x8433046	github.com/pion/webrtc/v2.(*PeerConnection).drainSRTP.func3+0x86	/go/src/github.com/pion/webrtc/peerconnection.go:983

2 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c77c 0x82be090 0x8432dc7 0x809e321
#	0x82be08f	github.com/pion/srtp.(*SessionSRTP).AcceptStream+0x3f			/go/pkg/mod/github.com/pion/srtp@v1.2.7/session_srtp.go:77
#	0x8432dc6	github.com/pion/webrtc/v2.(*PeerConnection).drainSRTP.func2+0x46	/go/src/github.com/pion/webrtc/peerconnection.go:963

2 @ 0x80758a4 0x8070524 0x806f9fb 0x80be807 0x80bf396 0x80bf37b 0x81bb84f 0x81cf69f 0x81ce467 0x82ede70 0x809e321
#	0x806f9fa	internal/poll.runtime_pollWait+0x4a			/usr/local/go/src/runtime/netpoll.go:184
#	0x80be806	internal/poll.(*pollDesc).wait+0x36			/usr/local/go/src/internal/poll/fd_poll_runtime.go:87
#	0x80bf395	internal/poll.(*pollDesc).waitRead+0x165		/usr/local/go/src/internal/poll/fd_poll_runtime.go:92
#	0x80bf37a	internal/poll.(*FD).ReadFrom+0x14a			/usr/local/go/src/internal/poll/fd_unix.go:219
#	0x81bb84e	net.(*netFD).readFrom+0x3e				/usr/local/go/src/net/fd_unix.go:208
#	0x81cf69e	net.(*UDPConn).readFrom+0x3e				/usr/local/go/src/net/udpsock_posix.go:47
#	0x81ce466	net.(*UDPConn).ReadFrom+0x46				/usr/local/go/src/net/udpsock.go:121
#	0x82ede6f	github.com/pion/ice.(*candidateBase).recvLoop+0xef	/go/pkg/mod/github.com/pion/ice@v0.7.10/candidate_base.go:92

2 @ 0x80758a4 0x8083e59 0x81fb473 0x809e321
#	0x81fb472	github.com/pion/sctp.(*Association).writeLoop+0x122	/go/pkg/mod/github.com/pion/sctp@v1.7.6/association.go:456

2 @ 0x80758a4 0x8083e59 0x829907f 0x81fb19c 0x809e321
#	0x829907e	github.com/pion/dtls/v2.(*Conn).Read+0xae		/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.7/conn.go:285
#	0x81fb19b	github.com/pion/sctp.(*Association).readLoop+0x13b	/go/pkg/mod/github.com/pion/sctp@v1.7.6/association.go:421

2 @ 0x80758a4 0x8083e59 0x82f85af 0x809e321
#	0x82f85ae	github.com/pion/ice.(*Agent).startConnectivityChecks.func1.1+0xae	/go/pkg/mod/github.com/pion/ice@v0.7.10/agent.go:612

1 @ 0x80758a4 0x804ca12 0x804c9ed 0x804c75c 0x8110f5b 0x8113daa 0x8110c2a 0x8111fcd 0x811127b 0x844e174 0x80754d2 0x809e321
#	0x8110f5a	testing.(*T).Run+0x2ca		/usr/local/go/src/testing/testing.go:961
#	0x8113da9	testing.runTests.func1+0x59	/usr/local/go/src/testing/testing.go:1202
#	0x8110c29	testing.tRunner+0x99		/usr/local/go/src/testing/testing.go:909
#	0x8111fcc	testing.runTests+0x22c		/usr/local/go/src/testing/testing.go:1200
#	0x811127a	testing.(*M).Run+0x13a		/usr/local/go/src/testing/testing.go:1117
#	0x844e173	main.main+0x103			_testmain.go:324
#	0x80754d1	runtime.main+0x201		/usr/local/go/src/runtime/proc.go:203

1 @ 0x80758a4 0x8083e59 0x84212aa 0x8422475 0x8110c2a 0x809e321
#	0x84212a9	github.com/pion/webrtc/v2.sendVideoUntilDone+0x129		/go/src/github.com/pion/webrtc/peerconnection_renegotation_test.go:19
#	0x8422474	github.com/pion/webrtc/v2.TestPeerConnection_RoleSwitch+0x434	/go/src/github.com/pion/webrtc/peerconnection_renegotation_test.go:187
#	0x8110c29	testing.tRunner+0x99						/usr/local/go/src/testing/testing.go:909

1 @ 0x814626c 0x81460da 0x8143133 0x83eea37 0x809e321
#	0x814626b	runtime/pprof.writeRuntimeProfile+0x7b			/usr/local/go/src/runtime/pprof/pprof.go:708
#	0x81460d9	runtime/pprof.writeGoroutine+0x79			/usr/local/go/src/runtime/pprof/pprof.go:670
#	0x8143132	runtime/pprof.(*Profile).WriteTo+0x2c2			/usr/local/go/src/runtime/pprof/pprof.go:329
#	0x83eea36	github.com/pion/transport/test.TimeOut.func1+0x56	/go/pkg/mod/github.com/pion/transport@v0.9.2/test/util.go:18

panic: timeout

goroutine 1691 [running]:
github.com/pion/transport/test.TimeOut.func1()
	/go/pkg/mod/github.com/pion/transport@v0.9.2/test/util.go:21 +0xda
created by time.goFunc
	/usr/local/go/src/time/sleep.go:168 +0x37
FAIL	github.com/pion/webrtc/v2	90.172s
```

**Comment by at-wat** [comment]
It is reproduced locally by increasing CPU load.

**Comment by at-wat** [comment]
```
openssl speed -multi 16
```
```
go test -v -race . -run TestPeerConnection_RoleSwitch -count 1000 -failfast
```

**Comment by at-wat** [comment]
As the order of function execution of multiple goroutine is not guaranteed, `startRenegotation` can be called before `startTransports`.

**Comment by at-wat** [comment]
@Sean-Der Still there is an another race condition related to this and https://github.com/pion/webrtc/issues/1039.
https://api.travis-ci.org/v3/job/665723204/log.txt
```
=== RUN   TestInvalidFingerprintCausesFailed
--- FAIL: TestInvalidFingerprintCausesFailed (10.42s)
    util.go:38: Unexpected routines on test end: 
        goroutine 1321 [select]:
        github.com/pion/dtls/v2.(*handshakeFSM).finish(0xa5ceee0, 0x860bb60, 0xa5cef20, 0x860d1e0, 0xa6601a0, 0x3, 0x0, 0x0)
        	/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.8/handshaker.go:288 +0x19c
        github.com/pion/dtls/v2.(*handshakeFSM).Run(0xa5ceee0, 0x860bb60, 0xa5cef20, 0x860d1e0, 0xa6601a0, 0xa83df01, 0x0, 0x0)
        	/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.8/handshaker.go:159 +0x2bf
        github.com/pion/dtls/v2.(*Conn).handshake.func2(0xa6601a0, 0x860bb60, 0xa5cef20, 0x1, 0xa656a40)
        	/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.8/conn.go:838 +0x88
        created by github.com/pion/dtls/v2.(*Conn).handshake
        	/go/pkg/mod/github.com/pion/dtls/v2@v2.0.0-rc.8/conn.go:836 +0x228
```
Close can be called before `t.conn = dtlsConn`
https://github.com/pion/webrtc/blob/4cb3b1a7c66816d96b4d385c9dd8675c59efdd00/dtlstransport.go#L280-L295

I'm still thinking that it's better to take lock during `dtls.Client` and `dtls.Server`, and the current implementation is routine unsafe.

**Comment by Sean-Der** [comment]
:+1: all for that change! sorry we didn't do it sooner. I assumed the only bug was the poor implementation around re-negotiation.

One thing that does concern me is that that `pion/dtls` might not be respecting nextConn being closed. When `Close` is called we close the ICE agent, this should flow downstream to DTLS. It is strange to me that it is deadlocking.

This might be effecting others users of `pion/dtls`, if upstream closes on them during a handshake it might keep trying?

**Comment by at-wat** [comment]
In this race condition, `DTLSTransport` just skips calling `dtlsConn.Close()` as it is not yet set, so there is no deadlock.

**Comment by Sean-Der** [comment]
Is `ICETransport` is closed? Originally I had thought [this](https://github.com/pion/dtls/blob/master/conn.go#L583) would return an error, then we can depend on that tearing down the `*dtls.Conn`

**Comment by at-wat** [comment]
I haven't checked `ICETransport` condition.

The problem is happens in the following sequence
1. main routine: `PeerConnection.startTransports()` is called
    - start sub routine
2. sub routine: `DTLSTransport.Start()` is called
3. main routine: `PeerConnection.startTransports()` ends
4. sub routine: `dtls.Client()` or `dtls.Server()` is called
5. main routine: `PeerConnection.Close()` is called
    - **`t.conn.Close()` is skipped as it is `nil`**
6. sub routine: `dtls.Client()` or `dtls.Server()` ends and **`t.conn = dtlsConn`, it will never closed**

**Comment by at-wat** [comment]
Logically, `PeerConnection` and `DTLSTransport` is causing this problem. It should be independent from `ICETransport`.

**Comment by Sean-Der** [comment]
The `dtlsEndpoint` *should* be closed though https://github.com/pion/webrtc/blob/master/dtlstransport.go#L281 so that means anything reading from it will fail.

So I would hope by [closing the iceTransport(https://github.com/pion/webrtc/blob/master/peerconnection.go#L1385-L1389) that `dtls.Client()` or `dtls.Server()` would then return. Internally the `dtlsEndpoint` that was given to `pion/dtls` is now going to cause everything to tear down.

If we don't properly tear down when `dtlsEndpoint` is closed this could be effecting other users of `pion/dtls`. They could have a dtls Client/Server that is reading from a Closed `net.Conn`, so the handshake should be stopped.

**Comment by at-wat** [comment]
Ah, I got what you concerned.
In short, it is caused because `ice.Conn` is not compatible with `net.Conn` at now.

`dtls` now expect that underlying `Conn` implements `SetReadDeadline` to make implementation of the asynchronous operations clean.
`ice.Conn` currently doesn't properly implement `SetReadDeadline`.
I recently added full spec `SetReadDeadline` to `transport/packetio.Buffer`, so it's very easy to support.
